bot/utils/scheduling.py

```diff
@@ -29,46 +31,58 @@ def __init__(self):
-        if task_id in self.scheduled_tasks:
+        log.trace(f"{self.cog_name}: scheduling task #{task_id}...")
+
+        if task_id in self._scheduled_tasks:
             log.debug(
                 f"{self.cog_name}: did not schedule task #{task_id}; task was already scheduled."
             )
             return
 
-        task: asyncio.Task = create_task(loop, self._scheduled_task(task_data))
+        task = asyncio.create_task(self._scheduled_task(task_data))
+        task.add_done_callback(partial(self._task_done_callback, task_id))
 
-        self.scheduled_tasks[task_id] = task
-        log.debug(f"{self.cog_name}: scheduled task #{task_id}.")
+        self._scheduled_tasks[task_id] = task
+        log.debug(f"{self.cog_name}: scheduled task #{task_id} {id(task)}.")
 
     def cancel_task(self, task_id: str) -> None:
         """Un-schedules a task."""
-        task = self.scheduled_tasks.get(task_id)
+        log.trace(f"{self.cog_name}: cancelling task #{task_id}...")
+
+        task = self._scheduled_tasks.get(task_id)
 
         if task is None:
             log.warning(f"{self.cog_name}: Failed to unschedule {task_id} (no task found).")
             return
 
         task.cancel()
-        log.debug(f"{self.cog_name}: unscheduled task #{task_id}.")
-        del self.scheduled_tasks[task_id]
-
-
-def create_task(loop: asyncio.AbstractEventLoop, coro_or_future: Union[Coroutine, asyncio.Future]) -> asyncio.Task:
-    """Creates an asyncio.Task object from a coroutine or future object."""
-    task: asyncio.Task = asyncio.ensure_future(coro_or_future, loop=loop)
-
-    # Silently ignore exceptions in a callback (handles the CancelledError nonsense)
-    task.add_done_callback(_silent_exception)
-    return task
+        log.debug(f"{self.cog_name}: unscheduled task #{task_id} {id(task)}.")
+        del self._scheduled_tasks[task_id]
 
+    def _task_done_callback(self, task_id: str, task: asyncio.Task) -> None:
+        """
+        Unschedule the task and raise its exception if one exists.
 
-def _silent_exception(future: asyncio.Future) -> None:
-    """Suppress future's exception."""
-    with contextlib.suppress(Exception):
-        future.exception()
+        If the task was cancelled, the CancelledError is retrieved and suppressed. In this case,
+        the task is already assumed to have been unscheduled.
+        """
+        log.trace(f"{self.cog_name}: performing done callback for task #{task_id} {id(task)}")
+
+        if task.cancelled():
+            with contextlib.suppress(asyncio.CancelledError):
+                task.exception()
+        else:
+            # Check if it exists to avoid logging a warning.
+            if task_id in self._scheduled_tasks:
+                # Only cancel if the task is not cancelled to avoid a race condition when a new
+                # task is scheduled using the same ID. Reminders do this when re-scheduling after
+                # editing.
+                self.cancel_task(task_id)
```

Comment:
Is there any chance of a task ending with another type of exception than `CancelledError` and a new task being rescheduled with the same `id` before this callback is executed? (Callbacks are scheduled with "call_soon", so they don't happen directly after the task is done, although they should happen "soon".)

If so, I think this will cancel the new task as we're cancelling it purely using the `task_id` which, if I read this comment right, may or may not assigned to the task this callback was initiated for.

If that's not realistic or this race condition is just a figment of my imagination (which it probably is), then just ignore this. Otherwise, an option like `if self._scheduled_tasks.get(task_id) is task` could work instead. That not only guarantees the `task_id` still exists but also that it's bound to the same task.

(On a side note, you're purely calling `cancel_task` to delete the reference in `scheduled_tasks` with additional logging right? There's no need to truly cancel it since we're inside of a `done` callback. The term "cancel" confused at first, because if the task were truly cancelled, the `task.exception()` that follows would propagate that `CancelledError`. However, after thinking about it, this can't happen since we're never actually cancelling a task here; we're just cleaning up our scheduled task references.)

Reply:
You've brought up some good points I think. I'm going to try to rework the task cancellation.

Reply:
OK, I came up with a much better and simpler solution. I don't know what the hell I was thinking before but the previously implementation was just wrong in several ways. See commit messages for details.